### PR TITLE
artifact/template: make destination path absolute inside taskdir

### PR DESCRIPTION
**Pull request by tgross**
Fixes https://github.com/hashicorp/nomad/issues/9148

Prior to Nomad 0.12.5, you could use `${NOMAD_SECRETS_DIR}/mysecret.txt` as
the `artifact.destination` and `template.destination` because we would always
append the destination to the task working directory. In the recent security
patch we treated the `destination` absolute path as valid if it didn't escape
the working directory, but this breaks backwards compatibility and
interpolation of `destination` fields.

This changeset partially reverts the behavior so that we always append the
destination, but we also perform the escape check on that new destination
after interpolation so the security hole is closed.

### CODE REVIEW

File: helper/funcs_test.go [hunk 1]
```diff
@@ -245,70 +246,104 @@ func TestCheckNamespaceScope(t *testing.T) {
...
 		t.Run(tc.name, func(t *testing.T) {
 			caseMsg := fmt.Sprintf("path: %v\ndir: %v", tc.path, tc.dir)
-			escapes, err := GetPathInSandbox(tc.dir, tc.path)
-			if tc.expectedErr != "" {
-				require.EqualError(t, err, tc.expectedErr, caseMsg)
-			} else {
-				require.NoError(t, err, caseMsg)
+
+			path := tc.path
+			if tc.join {
```
Comment (krishicks) [inline helper/funcs_test.go:342]: I'm finding the `if tc.join` here makes it hard to follow, because the caller to `PathEscapesSandbox` (with this change) _always_ calls `filepath.Join` on `config.TaskDir` and `src`. It might be clearer in the tests if the `path` included the `dir`, with the `tc.join` stuff removed entirely.

Comment (tgross) [reply helper/funcs_test.go:342]: I have this here because there are 3 callers: the two that unconditionally join the paths and the one in template that only joins if the `source` is absolute ([ref](https://github.com/hashicorp/nomad/pull/9149/files#diff-796a6e8da6dbce4e38ef692cf9f9e70b5b5770f9cf367e3e13ca1f86de356d01R556))

**Comment by tgross** [comment]
Relevant E2E tests against this PR:

```
$ go test -v . -suite VaultSecrets
=== RUN   TestE2E
...
=== RUN   TestE2E/VaultSecrets
=== RUN   TestE2E/VaultSecrets/*vaultsecrets.VaultSecretsTest
=== RUN   TestE2E/VaultSecrets/*vaultsecrets.VaultSecretsTest/TestVaultSecrets
...
--- PASS: TestE2E (75.05s)
    --- PASS: TestE2E/VaultSecrets (75.05s)
        --- PASS: TestE2E/VaultSecrets/*vaultsecrets.VaultSecretsTest (74.96s)
            --- PASS: TestE2E/VaultSecrets/*vaultsecrets.VaultSecretsTest/TestVaultSecrets (74.79s)
PASS
ok      github.com/hashicorp/nomad/e2e  75.178s
```

```
$ go test -v . -suite ConsulTemplate
=== RUN   TestE2E
...
=== RUN   TestE2E/ConsulTemplate
=== RUN   TestE2E/ConsulTemplate/*consultemplate.ConsulTemplateTest
=== RUN   TestE2E/ConsulTemplate/*consultemplate.ConsulTemplateTest/TestTemplateUpdateTriggers
...
--- PASS: TestE2E (48.58s)
    --- PASS: TestE2E/ConsulTemplate (48.58s)
        --- PASS: TestE2E/ConsulTemplate/*consultemplate.ConsulTemplateTest (48.49s)
            --- PASS: TestE2E/ConsulTemplate/*consultemplate.ConsulTemplateTest/TestTemplateUpdateTriggers (48.32s)

PASS
ok      github.com/hashicorp/nomad/e2e  48.702s
```

File: helper/funcs_test.go (hunk 1 above)
Comment (tgross) [reply helper/funcs_test.go:342]: Technically whether the path is joined doesn't matter to the `PathEscapesSandbox` at all, so maybe it'd be better just to have the paths "pre-joined" for clarity where appropriate?

Comment (krishicks) [reply helper/funcs_test.go:342]: The pre-joining is what I was looking for; it seems to exist on some tests already (the original ones?).

Comment (tgross) [reply helper/funcs_test.go]: Done in 2505858

**Review by schmichael — APPROVED** [review]

File: helper/funcs.go [hunk 2]
```diff
@@ -514,21 +514,16 @@ func CheckNamespaceScope(provided string, requested []string) []string {
...
-// joined to the sandbox directory. Returns an error if the path escapes the
-// sandbox directory.
-func GetPathInSandbox(sandboxDir, path string) (string, error) {
-	if !filepath.IsAbs(path) {
-		path = filepath.Join(sandboxDir, path)
-	}
+// PathEscapesSandbox returns whether a cleaned path inside the sandbox
+// directory (typically this will be the allocation directory) escapes.
+func PathEscapesSandbox(sandboxDir, path string) bool {
 	path = filepath.Clean(path)
```
[the remark this answers is not in the document]
Comment (tgross) [reply helper/funcs.go:520]: Good point, there's one case where we don't use `filepath.Join`, so I'll move the clean out to there rather than having it here.

[the remark this answers is not in the document]
Comment (tgross) [reply helper/funcs.go]: Done in 54109b4

**Comment by krishicks** [comment]
I wanted to follow-up and say the updated tests are way more clear to me. Thanks!
